File: Barrel/BLFileDownloader.m
/*
 Copyright (c) 2013, Barrel Team
 
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:
 * Redistributions of source code must retain the above copyright
 notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright
 notice, this list of conditions and the following disclaimer in the
 documentation and/or other materials provided with the distribution.
 * Neither the name of the Barrel Team nor the
 names of its contributors may be used to endorse or promote products
 derived from this software without specific prior written permission.
 
 THIS SOFTWARE IS PROVIDED BY Barrel Team ''AS IS'' AND ANY
 EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL Barrel Team BE LIABLE FOR ANY
 DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "BLFileDownloader.h"
#import "OEHUDAlert+DefaultAlertsAdditions.h"

@implementation BLFileDownloader

- (id)init {
    self = [super init];
    return self;
}

- (id)initWithProgressBar:(OEHUDProgressbar *)progressBar saveToPath:(NSString *)path {
    self = [super init];
    
    if (self) {
        [self setProgress:progressBar];
        [self setSavePath:path];
    }
    
    return self;
}

- (void)downloadWithNSURLConnectionFromURL:(NSString *)currentURL withCompletionBlock:(void (^)(int, NSString *))completionBlock {
    [self setCurrentURL:currentURL];
    NSURL *url = [NSURL URLWithString:currentURL];
    NSURLRequest *theRequest = [NSURLRequest requestWithURL:url cachePolicy:NSURLRequestReloadIgnoringLocalCacheData timeoutInterval:60];
    
    [self setReceivedData:[[NSMutableData alloc] initWithLength:0]];
    [self setCompletionBlock:[completionBlock copy]];
    
    NSURLConnection *connection = [[NSURLConnection alloc] initWithRequest:theRequest delegate:self startImmediately:NO];
    if (connection) {
        [self setConnection:connection];
    }
    else {
        OEHUDAlert *alert = [OEHUDAlert alertWithMessageText:@"Error connecting to server!" defaultButton:@"OK" alternateButton:@""];
        [alert runModal];
    }
}

- (void)startDownload {
    [[self connection] start];
}

- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
    [[self receivedData] setLength:0];
    [self setTotalBytes:[response expectedContentLength]];
    [[self progress] setMaxValue:(float)[self totalBytes]];
    [[self progress] setMinValue:0.0f];
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
    [[self receivedData] appendData:data];
    float progressive = (float)[[self receivedData] length];
    [[self progress] setValue:progressive];
}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error {
    OEHUDAlert *alert = [OEHUDAlert alertWithError:error];
    [alert runModal];
}

- (NSCachedURLResponse *)connection:(NSURLConnection *)connection willCacheResponse: (NSCachedURLResponse *)cachedResponse {
    return nil;
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
    NSLog(@"Succeeded! Received %ld bytes of data",(unsigned long)[[self receivedData] length]);
    NSError * error = nil;
    [[NSFileManager defaultManager] createDirectoryAtPath:[self savePath]
                              withIntermediateDirectories:YES
                                               attributes:nil
                                                    error:&error];
    if (error != nil) {
        NSLog(@"error creating directory: %@", error);
    }
    
    NSString *pathWithFilename = [NSString stringWithFormat:@"%@/%@", [self savePath], [[self currentURL] lastPathComponent]];
    [[self receivedData] writeToFile:pathWithFilename atomically:YES];
    NSLog(@"File written to path: %@", pathWithFilename);
    [self completionBlock](1, pathWithFilename);
}

@end

File: BarrelApp/BarrelApp/Classes/SubClasses/Network/BLFileDownloader.swift
//
//  BLFileDownloader.swift
//  BarrelBundle
//
//  Created by Thanos Siopoudis on 26/07/2015.
//  Copyright (c) 2015 AppCake Limited. All rights reserved.
//

import Foundation

class BLFileDownloader: NSObject, NSURLConnectionDataDelegate {
    var currentURL:NSURL?
    var receivedData:NSMutableData;
    var completionBlock:((Int, NSURL) -> Void)?
    var connection:NSURLConnection?
    var totalBytes:Int64 = 0;
    var saveURL:NSURL?
    
    override init() {
        self.receivedData = NSMutableData(length: 0)!;
        
        super.init();
    }
    
    convenience init(saveURL:NSURL?) {
        self.init();
        self.saveURL = saveURL;
    }
    
    func downloadWithNSURLConnectionFromURL(currentURL:NSURL?, completionBlock:((Int, NSURL) -> Void)!) {
        self.currentURL = currentURL;
        let theRequest:NSURLRequest = NSURLRequest(URL: self.currentURL!, cachePolicy: NSURLRequestCachePolicy.ReloadIgnoringLocalCacheData, timeoutInterval: 60);
        self.completionBlock = completionBlock;
        
        if let connection:NSURLConnection = NSURLConnection(request: theRequest, delegate: self, startImmediately: false) {
            self.connection = connection;
        }
        else {
            let errorAlert:NSAlert = NSAlert();
            errorAlert.addButtonWithTitle("OK");
            errorAlert.messageText = "Connection Error";
            errorAlert.informativeText = "Barrel failed to connect to the server. Please check your connection, or try again later.";
            errorAlert.alertStyle = NSAlertStyle.WarningAlertStyle;
            errorAlert.runModal();
        }
    }
    
    func startDownload() {
        self.connection!.start();
    }
    
    func cancelDownload() {
        self.connection!.cancel();
    }
    
    func connection(connection: NSURLConnection, didReceiveResponse response: NSURLResponse) {
        self.receivedData.length = 0;
        self.totalBytes = response.expectedContentLength;
    }
    
    func connection(connection: NSURLConnection, didReceiveData data: NSData) {
        self.receivedData.appendData(data);
    }
    
    func connection(connection: NSURLConnection, didFailWithError error: NSError) {
        let errorAlert:NSAlert = NSAlert(error: error);
        errorAlert.runModal();
    }
    
    func connection(connection: NSURLConnection, willCacheResponse cachedResponse: NSCachedURLResponse) -> NSCachedURLResponse? {
        return nil;
    }
    
    func connectionDidFinishLoading(connection: NSURLConnection) {
        if let sURL = self.saveURL {
            do {
                try NSFileManager.defaultManager().createDirectoryAtURL(sURL,
                    withIntermediateDirectories: true,
                    attributes: nil);
            }
            catch let err as NSError {
                let errorAlert:NSAlert = NSAlert(error: err);
                errorAlert.runModal();
            }
            
            let urlPathWithFilename:NSURL = sURL.URLByAppendingPathComponent(self.currentURL!.lastPathComponent!);
            self.receivedData.writeToURL(urlPathWithFilename, atomically: true);
            if let cBlock = self.completionBlock {
                cBlock(1, urlPathWithFilename);
            }
        }
    }
}
















